cranelift-codegen/src/simple_preopt.rs

```diff
@@ -516,6 +517,27 @@ fn simplify(pos: &mut FuncCursor, inst: Inst) {
                     imm,
                 } = pos.func.dfg[iconst_inst]
                 {
+                    // For signed comparisons, we may need to cast the immediate to the instruction
+                    // result (controlling) type.
+                    let imm = match cond {
+                        IntCC::SignedLessThan
+                        | IntCC::SignedLessThanOrEqual
+                        | IntCC::SignedGreaterThan
+                        | IntCC::SignedGreaterThanOrEqual => {
+                            let val: i64 = imm.into();
+                            let new_imm = match pos.func.dfg.ctrl_typevar(inst) {
+                                I8 => (val as i8) as i64,
```

Comment:
nit: I am tempted to suggest to move this code either as `impl From<(ir::Type, i64)> for i64` or as a function on the `ir::Type` named `Type::sign_extend(i64) -> i64`.

However, I do not yet know how frequent this sign-extension might be. So feel free to discard this comment if there is no other instances.